File: apps/desktop/src/lib/commands.ts
// Typed wrappers for the Tauri commands in src-tauri/src/commands.rs.

import { invoke } from "@tauri-apps/api/core";
import { openUrl } from "@tauri-apps/plugin-opener";
import { listen, type UnlistenFn } from "@tauri-apps/api/event";

import type { DocumentState } from "../bindings/DocumentState";
import type { FileEntry } from "../bindings/FileEntry";
import type { Fusen } from "../bindings/Fusen";
import type { NewAnchor } from "../bindings/NewAnchor";
import type { SessionInfo } from "../bindings/SessionInfo";
import type { Settings } from "../bindings/Settings";
import type { Status } from "../bindings/Status";

export const getSession = () => invoke<SessionInfo | null>("get_session");

export const listFiles = () => invoke<FileEntry[]>("list_files");

/** Checks that a link target is a Markdown file in the project. */
export const resolveDocument = (path: string) => invoke<string>("resolve_document", { path });

/** Opens a URL in the default browser. */
export const openExternal = (url: string) =>
  openUrl(url).catch(() => {
    // Running in a browser with the mock backend.
    window.open(url, "_blank");
  });

export const readDocument = (path: string) =>
  invoke<DocumentState>("read_document", { path });

export const addFusen = (args: {
  path: string;
  docHash: string;
  kind: string;
  body: string;
  anchor: NewAnchor | null;
}) => invoke<Fusen>("add_fusen", args);

export const editFusen = (path: string, id: string, kind: string, body: string) =>
  invoke<void>("edit_fusen", { path, id, kind, body });

export const deleteFusen = (path: string, ids: string[]) =>
  invoke<void>("delete_fusen", { path, ids });

export const setStatus = (path: string, ids: string[], status: Status) =>
  invoke<void>("set_status", { path, ids, status });

export const addReply = (path: string, id: string, body: string) =>
  invoke<void>("add_reply", { path, id, body });

export const getSettings = () => invoke<Settings>("get_settings");

export const defaultSettings = () => invoke<Settings>("default_settings");

export const saveSettings = (settings: Settings) =>
  invoke<void>("save_settings", { settings });

// Events sent from src-tauri (see watcher.rs and lib.rs).
export const onOpenTarget = (f: (info: SessionInfo) => void): Promise<UnlistenFn> =>
  listen<SessionInfo>("open-target", (e) => f(e.payload));

export const onFilesChanged = (f: () => void): Promise<UnlistenFn> =>
  listen("files-changed", () => f());

export const onDocumentChanged = (f: (path: string) => void): Promise<UnlistenFn> =>
  listen<string>("document-changed", (e) => f(e.payload));

export const onShowSettings = (f: () => void): Promise<UnlistenFn> =>
  listen("show-settings", () => f());

export const onImagesChanged = (f: () => void): Promise<UnlistenFn> =>
  listen("images-changed", () => f());

export const onSettingsChanged = (f: () => void): Promise<UnlistenFn> =>
  listen("settings-changed", () => f());

File: apps/desktop/src-tauri/src/commands.rs
//! Tauri commands exposed to the UI. They only call fusen-core.

use std::sync::Mutex;

use fusen_core::config::{Config, Settings};
use fusen_core::fusen::ops::{self, DocumentState, NewAnchor};
use fusen_core::fusen::{Fusen, Status};
use fusen_core::project::Project;
use serde::Serialize;
use tauri::State;
use ts_rs::TS;

use crate::session::{Session, SessionInfo};

pub type AppState = Mutex<Option<Session>>;

type Result<T> = std::result::Result<T, String>;

#[derive(Debug, Clone, Serialize, TS)]
#[serde(rename_all = "camelCase")]
#[ts(export)]
pub struct FileEntry {
    pub path: String,
    pub open_count: u32,
    pub outdated_count: u32,
}

fn with_project<T>(
    state: &State<AppState>,
    f: impl FnOnce(&Project, &Session) -> fusen_core::Result<T>,
) -> Result<T> {
    let guard = state.lock().unwrap();
    let session = guard.as_ref().ok_or("no project is open")?;
    f(&session.project, session).map_err(|e| e.to_string())
}

/// Resolves a document path from the UI, making sure it is inside the project.
fn doc(project: &Project, path: &str) -> fusen_core::Result<String> {
    project.resolve_file(path)
}

#[tauri::command]
pub fn get_session(state: State<AppState>) -> Option<SessionInfo> {
    state.lock().unwrap().as_ref().map(Session::info)
}

#[tauri::command]
pub fn list_files(state: State<AppState>) -> Result<Vec<FileEntry>> {
    with_project(&state, |project, session| {
        Ok(project
            .markdown_files(&session.target)
            .into_iter()
            .map(|path| {
                // An invalid fusen file should not hide the whole list; its
                // error is shown when the document is opened.
                let (open, outdated) = ops::state_counts(project, &path).unwrap_or((0, 0));
                FileEntry {
                    path,
                    open_count: open as u32,
                    outdated_count: outdated as u32,
                }
            })
            .collect())
    })
}

/// Checks that a link target is a Markdown file in the project, and returns
/// its path relative to the project root.
#[tauri::command]
pub fn resolve_document(state: State<AppState>, path: String) -> Result<String> {
    with_project(&state, |project, _| doc(project, &path))
}

#[tauri::command]
pub fn read_document(state: State<AppState>, path: String) -> Result<DocumentState> {
    with_project(&state, |project, _| {
        ops::read_document(project, &doc(project, &path)?)
    })
}

#[tauri::command]
pub fn add_fusen(
    state: State<AppState>,
    path: String,
    doc_hash: String,
    kind: String,
    body: String,
    anchor: Option<NewAnchor>,
) -> Result<Fusen> {
    with_project(&state, |project, _| {
        ops::add_fusen(
            project,
            &doc(project, &path)?,
            Some(doc_hash),
            &kind,
            &body,
            anchor,
        )
    })
}

#[tauri::command]
pub fn edit_fusen(
    state: State<AppState>,
    path: String,
    id: String,
    kind: String,
    body: String,
) -> Result<()> {
    with_project(&state, |project, _| {
        ops::edit_fusen(project, &doc(project, &path)?, &id, &kind, &body)
    })
}

#[tauri::command]
pub fn delete_fusen(state: State<AppState>, path: String, ids: Vec<String>) -> Result<()> {
    with_project(&state, |project, _| {
        ops::delete_fusen(project, &doc(project, &path)?, &ids)
    })
}

#[tauri::command]
pub fn set_status(
    state: State<AppState>,
    path: String,
    ids: Vec<String>,
    status: Status,
) -> Result<()> {
    // No IDs would mean every fusen on the document.
    if ids.is_empty() {
        return Ok(());
    }
    with_project(&state, |project, _| {
        for result in ops::set_status(project, &doc(project, &path)?, &ids, status)? {
            result?;
        }
        Ok(())
    })
}

#[tauri::command]
pub fn add_reply(state: State<AppState>, path: String, id: String, body: String) -> Result<()> {
    with_project(&state, |project, _| {
        let author = Config::load_default()?.user_name();
        ops::add_reply(project, &doc(project, &path)?, &id, &author, &body).map(|_| ())
    })
}

#[tauri::command]
pub fn get_settings() -> Result<Settings> {
    Config::load_default()
        .map(|c| c.settings())
        .map_err(|e| e.to_string())
}

/// The settings with every key at its default, for the Reset buttons.
#[tauri::command]
pub fn default_settings() -> Settings {
    Config::default().settings()
}

#[tauri::command]
pub fn save_settings(settings: Settings) -> Result<()> {
    let mut config = Config::load_default().map_err(|e| e.to_string())?;
    config.replace(&settings).map_err(|e| e.to_string())?;
    config.save().map_err(|e| e.to_string())
}
